Release notes for 1.2.0 release. (#3951)
dotnet/machinelearning: docs/release-notes/1.2.0/release-1.2.0.md

# [ML.NET](http://dot.net/ml) 1.2.0 
## **General Availability**
- **Microsoft.ML.TimeSeries**  
    - Anomaly detection algorithms (Spike and Change Point):
      - Independent and identically distributed.
      - Singular spectrum analysis.
      - Spectral residual from Azure Anomaly Detector/Kensho team.
    - Forecasting models:
      - Singular spectrum analysis.
    - Prediction Engine for online learning

>       - Allows updating time series model with new observations at scoring so that the user does not have re-train the time series with old data each time.

Reviewer: Instead of "allows", use "enables"

>       - Allows updating time series model with new observations at scoring so that the user does not have re-train the time series with old data each time.

Reviewer 2: small typo:
does not have *to* re-train

     [Samples](https://github.com/dotnet/machinelearning/tree/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/TimeSeries)

- **Microsoft.ML.OnnxTransformer**  

>    Allows scoring of ONNX models in the learning pipeline. Uses ONNX Runtime v0.4.

Reviewer: "Enables" instead of "allows"

      
   [Sample](https://github.com/dotnet/machinelearning/blob/master/docs/samples/Microsoft.ML.Samples/Dynamic/Transforms/ApplyOnnxModel.cs)

> - **Microsoft.ML.TensforFlow**  

Reviewer 3: Spelling:
```suggestion
- **Microsoft.ML.TensorFlow**  
```

>    Allows scoring of TensorFlow models in the learning pipeline. Uses TensforFlow v1.13. Very useful for image and text classification.

Reviewer: "Enables" instead of "allows"

Perhaps clarify why it is very useful for image/text classifications. Does it enable them easier? Is it a community asked feature.

>    Allows scoring of TensorFlow models in the learning pipeline. Uses TensforFlow v1.13. Very useful for image and text classification.

Reviewer 3: Spelling:
```suggestion
   Allows scoring of TensorFlow models in the learning pipeline. Uses TensorFlow v1.13. Very useful for image and text classification.
```

   [Samples](https://github.com/dotnet/machinelearning/tree/master/docs/samples/Microsoft.ML.Samples/Dynamic/TensorFlow)

## **New Features**

> - **Tree-based featurization** ([#3812](https://github.com/dotnet/machinelearning/pull/3812))

Reviewer 4: Can we provide reference or more technical details?

Reviewer 5: Here are two references.
- p. 9 at https://www.csie.ntu.edu.tw/~r01922136/kaggle-2014-criteo.pdf
- Section 3 at http://www.quinonero.net/Publications/predicting-clicks-facebook.pdf.
ML.NET's tree featurization trains a tree-based model and then maps input feature vector to several non-linear feature vectors. Those generated feature vectors are
- the leaves it falls into. It's a binary vector with ones happens at the indexes of reached leaves,
- the paths that the input vector passes before hitting the leaves, and
- the reached leaves values.

>   Generating features using tree structure has been a popular technique in data mining. This PR exposes this internal-only feature to the public. 

Reviewer 3: ```suggestion
  Generating features using tree structure has been a popular technique in data mining. Useful for capturing feature interactions when creating a stacked model, dimensionality reduction, or featurizing towards an alternative label. 
```

    Specifically it contains functionality for:
    - Dependency Injection
    - Pooling PredictionEngines
    - Reloading models when the file or URI has changed
    - Hooking ML.NET logging to Microsoft.Extensions.Logging

## **Bug Fixes**
### Serious
- **Time series Sequential Transform needs to have a binding mechanism:** This bug made it impossible to use time series in NimbusML. ([#3875](https://github.com/dotnet/machinelearning/pull/3875))

- **build errors resulting from upgrading to VS2019 compilers:** The default CMAKE_C_FLAG for debug configuration sets /ZI to generate a PDB capable of edit and continue. In the new compilers, this is incompatible with /guard:cf which we set for security reasons. ([#3894](https://github.com/dotnet/machinelearning/pull/3894)) 

- **LightGBM Evaluation metric parameters:** In LightGbm EvaluateMetricType where if a user specified EvaluateMetricType.Default, the metric would not get added to the options Dictionary, and LightGbmWrappedTraining would throw because of that. ([#3815](https://github.com/dotnet/machinelearning/pull/3815))

> - **Change default EvaluationMetric for LightGbm:** In [ML.NET](dot.net/ml), the default EvaluationMetric for LightGbm is set to EvaluateMetricType.Error for multiclass, EvaluationMetricType.LogLoss for binary, and so on. This leads to inconsistent behavior from the user's perspective. ([#3859](https://github.com/dotnet/machinelearning/pull/3859))

Reviewer: "etc." instead of "and so on"

### Other
- CustomGains should allow multiple values in argument attribute. ([#3854](https://github.com/dotnet/machinelearning/pull/3854))

## **Breaking Changes**
None

## **Enhancements**

> - Fixes the Hardcoded Sigmoid value from -.5 to the value specified during training. ([#3850](https://github.com/dotnet/machinelearning/pull/3850))

Reviewer 3: ```suggestion
- Fixes the Hardcoded Sigmoid value from -0.5 to the value specified during training. ([#3850](https://github.com/dotnet/machinelearning/pull/3850))
```

- Fix TextLoader constructor and add exception message. ([#3788](https://github.com/dotnet/machinelearning/pull/3788))
- Introduce the `FixZero` argument to the LogMeanVariance normalizer. ([#3916](https://github.com/dotnet/machinelearning/pull/3916))

> - Ensembles trainer now work with ITrainerEstimators instead of ITrainers, ([#3796](https://github.com/dotnet/machinelearning/pull/3796)) 

Reviewer 2: This is internal only, should we mention it at all?

- LightGBM Unbalanced Data Argument. ([#3925](https://github.com/dotnet/machinelearning/pull/3925))
- Tree based trainers implement ICanGetSummaryAsIDataView. ([#3892](https://github.com/dotnet/machinelearning/pull/3892))

> 

Reviewer 3: As we're doing a joint release:
```suggestion
## **CLI and AutoML API**
- Internationalization fixes to generate proper ML.NET C# code. ([#3725](https://github.com/dotnet/machinelearning/pull/3725))
- Automatic Cross Validation for small datasets, and CV stability fixes. ([#3794](https://github.com/dotnet/machinelearning/pull/3794))
- Code cleanup to match .NET style. ([#3823](https://github.com/dotnet/machinelearning/pull/3823))

```
```

## **Documentation and Samples**
- Samples for applying ONNX model to in-memory images. ([#3851](https://github.com/dotnet/machinelearning/pull/3851))
- Reformatted Ranking samples to width 85 so the horizonal scroll bar does not appear on docs webpage.([#3930](https://github.com/dotnet/machinelearning/pull/3930))

## **Remarks**

> - Github issue count decreased from ~550 to ~380.

Reviewer: Perhaps mention why it decreased, "we closed x issues dropping the count from .. to .."
